### ISSUE TITLE
Re-export types from monorepo internal dependencies

### ISSUE THREAD

**Issue by gxxcastillo** [qmhc/vite-plugin-dts#302]
### Description

I have a packageA in a monorepo and it depends on packageB and packageC, packageC depends on packageD.

When I run vite build on packageA using the dts plugin and `{ rollupTypes: true }`, I don't see the types from packageB, packageC, and packageD being exported from packageA.

### Suggested solution

Have a option that allows for types from dependencies to be included in the generated *.d.ts files

### Validations

- [X] Read the [FAQ](https://github.com/qmhc/vite-plugin-dts#faq).
- [X] Check that there isn't [already an issue](https://github.com/qmhc/vite-plugin-dts/issues) that reports the same bug to avoid creating a duplicate.

**Comment by qmhc** [comment]
If the depended relationships are correct, TypeScript may always resolve the declarations map when you are using them.

Like the following:

```ts
// A.d.ts
import D from 'packageD'

interface A extends D {
  // ...
}
```

**Comment by ohkimur** [comment]
@qmhc We need a fix for this. It's making life hard when using monorepos. For example, if there is at least one private package that is imported into another public package, the private import is not resolved for types currently. At least not consistently.

It seems that if we have `rollupTypes` set on true and `bundledPackages` set to ["@company/*"] then for some reason it works. However, if we have only the `bundledPackages` param set, the imports are not resolved again. After looking deep through the types and the code of `vite-plugin-dts` I think this is not as intended.

It would be great if we could tell to `vite-plugin-dts` which packages are bundled via `bundledPackages` for it to bundle the types, but without being required to rollup the types via `rollupTypes` param. I personally don't want to be forced to rollup my types, and also this makes life hard when working with packages that have multiple entry points.

My proposal is to just decouple them the right way so that we make this plugin even more better. Keep it up!

**Comment by qmhc** [comment]
@ohkimur You can view this demo: https://stackblitz.com/edit/vitejs-vite-oaq33v?file=vite.config.ts&view=editor.

If you think this is not match your case, you can edit it to reproduce you issue.

**Comment by avp1598** [comment]
@ohkimur for me even adding `rollupTypes` and `bundledPackages` does not work for some reason, it generates this `index.d.ts` file

```js
import { Something } from './common_package';
```

but it there is not `common_package` file, I am on turborepo yarn monrepo setup and i am trying to use a common package as an internal dependency on another package i am building using vite

**Comment by ohkimur** [comment]
@avp1598 Your setup is very similar to mine. I also use `turborepo` and multiple packages that are powered by `vite`.

Here is how I setup the `dts` for my packages:

```ts
    dts({
      exclude: ["tests"],
      outDir: "dist/types",
      rollupTypes: true,
      bundledPackages: ["@acme/*"],
    }),
```

**Comment by pandeymangg** [comment]
@ohkimur I'm also running into similar problems. I have a monorepo setup with turborepo and pnpm and it has a couple of packages built with vite. I have a `packageA` which uses an internal package `packageB` with multiple exports like `packageB/app` and `packageB/website`.

For some reason, having `bundledPackages` set to `bundledPackages: ["packageB"]` in the vite config of `packageA` does not inline the declarations of `packageB` and its the same problem that you're talking about. It should work though, since the package.json for `packageA` does have `packageB` as a dependency but its not 😞

**Comment by ohkimur** [comment]
@pandeymangg Yeah, this is also the case for my monorepo. You can `rollupTypes` as a workaround, but that is far from the correct behavior.

@qmhc Do you have any ideas on how to fix this? If you give at least some hints, I could make a PR to fix this one.

**Comment by qmhc** [comment]
@ohkimur Well, it's a difficult work. You need to analyze the relationship between packages, and manually find the modules and symbols in internal packages which are used in the public packages.

After some research I found that it even need to use internal APIs of TypeScript. 😓

**Comment by augusttty** [comment]
facing the same issue in monorepo, the dependencies may be a private module.it would be much better that` vite-plugin-dts` can extract some type declaration from dependencies into output...

**Comment by Joseph7451797** [comment]
I also encounter the issue. I have several packages in my monorepo, package A imports a shared type declaration from the internal package B. Package A will be built as a library with dts file distributed. In the distributed dts file, we still see the package B's path which will causes error if the developer installs this package.

My workaround is to re-create an identical type in package A and only use it in package A. And leverage [Vitest's type test](https://vitest.dev/guide/testing-types) to test if the re-created one is aligned with the original one in package B.

**Comment by qmhc** [comment]
Maybe you can try the following workaround:
- Build private package A first
- Copy the declaration files of A to the B's output target as `[B-dist]/packageA`
- Build public package B, use `beforeWriteFile` option and change the those `import xxx from 'packageA'` to `import xxx from '[relative to dist]/packageA'`

**Comment by tschoartschi** [comment]
This sounds very similar to the problems we are experiencing as well. We have a monorepo that somehow looks like:

```
my-app/
└── packages/
    ├── configurator
    |   ├── fileA.ts
    |   └── fileB.ts
    ├── common
    |   ├── fileC.ts
    |   └── fileD.ts
    └── embedding-lib
         ├── fileE.ts
         └── index.ts
```

Now I want to produce an `index.d.ts` file from `my-app/packages/embedding-lib/index.ts`. The strange thing is, that some definitions of some files work, for example if I import types from `my-app/packages/configurator/fileB.ts` inside `embedding-lib`, then it works and the definitions are inlined into the final `index.d.ts` but from other files not and only the relative path is taken. This of course breaks because the `my-app/packages/configurator` and `my-app/packages/common` will never be published anywhere. 

Is there some workaround? Does someone have an idea what could be the problem?

**Comment by gxxcastillo** [comment]
Apologies, I forgot to follow up on this post after finding my solutions.  Also, note, I'm not up to date on this thread so I don't know if there's an update to this library that solves this but this was my solution:

my solution back then was to create an api-extractor.json file in the root of my published package  and list all of my internal packages:

```
{
   bundledPackages: ['internal-package1', 'internal-package2', etc.. ]
}
```

[Here's the thread](https://rushstack.zulipchat.com/#narrow/stream/262521-api-extractor/topic/Error.3A.20Unable.20to.20determine.20semantic.20information) that lead me to the answer

**Comment by Gawdfrey** [comment]
> Apologies, I forgot to follow up on this post after finding my solutions. Also, note, I'm not up to date on this thread so I don't know if there's an update to this library that solves this but this was my solution:
> 
> Create an api-extractor.json file in the root of my published package and list all of my internal packages:
> 
> ```
> {
>    bundledPackages: ['internal-package1', 'internal-package2', etc.. ]
> }
> ```
> 
> [Here's the thread](https://rushstack.zulipchat.com/#narrow/stream/262521-api-extractor/topic/Error.3A.20Unable.20to.20determine.20semantic.20information) that lead me to the answer

Did you have to reference the file somehow? Adding the file does nothing for me, only thing that works is bundledPackages and rollupTypes together..

**Comment by Shuyinsama** [comment]
I am running into the same issue. I am using normal npm workspaces (no other managers or monorepo tools)

We have 2 packages that are being published and 1 internal package that houses all interfaces that the other 2 packages implement.

`packageA` and `packageB` both have `internalPackage` as a `devDependency`.

When building `packageA` and `packageB` we want the interfaces/types etc from `internalPackage` to be included in `packageA` and `packageB` without publishing `internalPackage`.

Both `packageA` and `packageB` have the following `vite.config.ts`

```
plugins: [
    dts({
      include: ['lib'],
      rollupTypes: true,
      bundledPackages: ['internalPackage'],
      tsconfigPath: resolve(__dirname, 'tsconfig.app.json'),
    }),
  ],
```

In both `packageA` and `packageB` we use the package as follows `"internalPackage": "*"`
What I noticed is that when the `package.json` of `internalPackage` is set to `private` it still needs the `version` property otherwise I think npm can't resolve the `*` to a version and thus can't find the `internalPackage` directory to actually include the interfaces and types.

I added the `version` tag to the `package.json` of `internalPackage` and after that the interfaces and types were included again in the `packageA` and `packageB`

**Comment by qmhc** [comment]
Maybe a workaround: https://stackblitz.com/edit/vitejs-vite-pwuvfelt?file=packages%2Ffoo%2Fvite.config.ts&view=editor

Key configuration: `resolve.alias` of Vite, `paths` of tsconfig.json, including internal packages for dts plugin option.
